test/apps/iodemo/io_demo.cc

```diff
@@ -67,40 +68,72 @@ typedef struct {
 template<class T>
 class MemoryPool {
 public:
-    MemoryPool(size_t buffer_size) :
+    MemoryPool(size_t buffer_size, size_t offcache) :
         _num_allocated(0), _buffer_size(buffer_size) {
+
+        for (size_t i = 0; i < offcache; ++i) {
+            _offcache_queue.push(get_free());
+        }
     }
 
     ~MemoryPool() {
-        if (_num_allocated != _freelist.size()) {
+        while (!_offcache_queue.empty()) {
+            put_free(_offcache_queue.front());
+            _offcache_queue.pop();
+        }
+
+        if (_num_allocated != _free_stack.size()) {
             LOG << "Some items were not freed. Total:" << _num_allocated
-                << ", current:" << _freelist.size() << ".";
+                << ", current:" << _free_stack.size() << ".";
+        }
+
+        for (size_t i = 0; i < _free_stack.size(); i++) {
+            delete _free_stack[i];
+        }
+    }
+
+    inline T* get() {
+        if (_offcache_queue.empty()) {
+            return get_free();
+        } else {
+            T* item = _offcache_queue.front();
+            _offcache_queue.pop();
+            _offcache_queue.push(get_free());
+            return item;
         }
+    }
 
-        for (size_t i = 0; i < _freelist.size(); i++) {
-            delete _freelist[i];
+    inline void put(T* item) {
+        if (_offcache_queue.empty()) {
+            put_free(item);
+        } else {
+            put_free(_offcache_queue.front());
+            _offcache_queue.pop();
+            _offcache_queue.push(item);
```

Comment:
we should always put back to freelist here (put_free)
if offcache_queue will need more buffers, it will pull from freelist